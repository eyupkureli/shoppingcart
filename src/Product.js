import React from 'react';
import {
    Card,
    CardBody,
    CardTitle,
    CardSubtitle,
    Button,
  } from "reactstrap";

const Product = (props) => {
    const {product, addProduct} = props;
    return (
        <div>
            <Card>
              <CardBody>
                <CardTitle tag="h5">{product.name}</CardTitle>
                <CardSubtitle tag="h6" className="mb-2 text-muted">
                  Â£ {product.price}
                </CardSubtitle>
                <Button data-testid = "button" label = "add your slot" onClick={() => addProduct(product)} color="success">Add Your Slot</Button>
              </CardBody>
            </Card>
            
        </div>
    )
}

export default Product;
