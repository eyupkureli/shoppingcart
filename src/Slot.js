import React from "react";
import { Button } from "reactstrap";

const Slot = (props) => {
  const { slotItems, removeProduct } = props;
  const productsPrice = slotItems.reduce(
    (accumulator, currentItem) =>
      accumulator + currentItem.price * currentItem.amount,
    0
  );

  

  let beanAmount = slotItems
    .filter((slot) => slot.id === 1)
    .reduce((accumulator, currentItem) => accumulator + currentItem.amount, 0);
  let beanDiscount = Math.floor(beanAmount / 3) * 0.5;

  let colaAmount = slotItems
    .filter((slot) => slot.id === 2)
    .reduce((accumulator, currentItem) => accumulator + currentItem.amount, 0);
  let colaDiscount = Math.floor(colaAmount / 2) * 0.4;

  return (
    <aside>
      <h1>Your Slot</h1>
      <div>
        {slotItems.length === 0 && (
          <div>
            <p>There is no product...</p>
          </div>
        )}
      </div>
      {slotItems.map((slotItem) => (
        <div key={slotItem.id}>
          <div>{slotItem.name}</div>
          <div>
            <Button data-testid = "remove" onClick={() => removeProduct(slotItem)} color="danger">
              Remove
            </Button>
          </div>
          <div>
            Quantity :{slotItem.amount}
          </div>
          <hr></hr>
        </div>
      ))}
      {slotItems.length !== 0 && (
        <>
         
          <div>
            <div>Total Price </div>
            <div>
              Â£ {(productsPrice - beanDiscount - colaDiscount).toFixed(2)}
            </div>
          </div>
        </>
      )}
    </aside>
  );
};

export default Slot;
