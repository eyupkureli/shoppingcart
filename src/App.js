import React, {useState} from "react";
import "./App.css";
import Navi from "./Navi";
import Shopping from "./Shopping";
import Slot from "./Slot";
import { Container, Row, Col } from "reactstrap";
import data from "./data";

function App() {
  const {products} = data;
  const [slotItems, setSlotItems] = useState([]);
  
  const addProduct = (product) => {
    const inSlot = slotItems.find(item => item.id === product.id);
    if(inSlot ) {
      setSlotItems(slotItems.map(item => item.id === product.id ? {...inSlot, amount: inSlot.amount + 1} : item
        ));
    }else {
      setSlotItems([...slotItems, {...product, amount: 1}]);
    }
  };

  const removeProduct = (product) => {
    const inSlot = slotItems.find(item => item.id === product.id);
    if(inSlot.amount === 1) {
      setSlotItems(slotItems.filter(item => item.id !== product.id));
    }else {
      setSlotItems(slotItems.map(item => item.id === product.id ? {...inSlot, amount: inSlot.amount - 1} : item
        ));
    }
  }

  return (
    <div>
      <Container>
        <Row>
          <Navi></Navi>
        </Row>
        <Row>
          <Col xs = "6"><Shopping addProduct = {addProduct} products = {products}></Shopping></Col>
          <Col xs = "6"><Slot addProduct = {addProduct} removeProduct = {removeProduct} slotItems = {slotItems}></Slot></Col>      
        </Row>
      </Container>
    </div>
  );
}

export default App;
