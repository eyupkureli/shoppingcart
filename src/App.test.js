import React from 'react';
import ReactDom from 'react-dom';
//import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { render, cleanup} from '@testing-library/react';
import App from './App';
import Product from './Product';
import Slot from './Slot';

afterEach(cleanup);

test('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDom.render(<App/>, div);
  ReactDom.unmountComponentAtNode(div);
});

test('renders button correctly', () => {
  const { getByTestId } = render(<Product/>);
  expect(getByTestId('button')).toHaveTextContent("Add Your Slot");
});

test('renders remove button correctly', () => {
  const { getByTestId } = render(<Slot/>);
  expect(getByTestId('remove')).toHaveTextContent("Remove");
});


test('matches the snapshot', () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});

