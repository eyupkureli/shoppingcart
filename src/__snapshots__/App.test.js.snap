// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div>
        <nav
          className="navbar navbar-expand-md navbar-light bg-light"
        >
          <a
            className="navbar-brand"
            href="/"
          >
            Eyup Supermarket
          </a>
          <button
            aria-label="Toggle navigation"
            className="navbar-toggler"
            onClick={[Function]}
            type="button"
          >
            <span
              className="navbar-toggler-icon"
            />
          </button>
          <div
            className="collapse navbar-collapse"
            style={Object {}}
          >
            <ul
              className="mr-auto navbar-nav"
            />
          </div>
        </nav>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-6"
      >
        <main>
          <h1>
            Products
          </h1>
          <div>
            <div>
              <div
                className="card"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-title"
                  >
                    Tin of Beans
                  </h5>
                  <h6
                    className="mb-2 text-muted card-subtitle"
                  >
                    £ 
                    0.5
                  </h6>
                  <button
                    aria-label={null}
                    className="btn btn-success"
                    data-testid="button"
                    label="add your slot"
                    onClick={[Function]}
                    type="button"
                  >
                    Add Your Slot
                  </button>
                </div>
              </div>
            </div>
            <div>
              <div
                className="card"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-title"
                  >
                    Can of Cola
                  </h5>
                  <h6
                    className="mb-2 text-muted card-subtitle"
                  >
                    £ 
                    0.7
                  </h6>
                  <button
                    aria-label={null}
                    className="btn btn-success"
                    data-testid="button"
                    label="add your slot"
                    onClick={[Function]}
                    type="button"
                  >
                    Add Your Slot
                  </button>
                </div>
              </div>
            </div>
            <div>
              <div
                className="card"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-title"
                  >
                    Oranges
                  </h5>
                  <h6
                    className="mb-2 text-muted card-subtitle"
                  >
                    £ 
                    1.99
                  </h6>
                  <button
                    aria-label={null}
                    className="btn btn-success"
                    data-testid="button"
                    label="add your slot"
                    onClick={[Function]}
                    type="button"
                  >
                    Add Your Slot
                  </button>
                </div>
              </div>
            </div>
          </div>
        </main>
      </div>
      <div
        className="col-6"
      >
        <aside>
          <h1>
            Your Slot
          </h1>
          <div>
            <div>
              <p>
                There is no product...
              </p>
            </div>
          </div>
        </aside>
      </div>
    </div>
  </div>
</div>
`;
